### PR TITLE
[react] add library managed attributes

### PR DESCRIPTION
**Pull request by ferdaber**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [x] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

Part 2 of this closed PR: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/27267. Dependent on https://github.com/DefinitelyTyped/DefinitelyTyped/pull/27378.

### CODE REVIEW

**Comment by weswigham** [comment]
Did y'all mean to open this on your own fork?

File: types/react/index.d.ts [hunk 1]
```diff
@@ -2282,6 +2281,11 @@ declare namespace React {
     }
 }
 
+type Defaultize<P, D = {}> =
+    & Pick<P, Exclude<keyof P, keyof D>>
+    & Partial<Pick<P, Extract<keyof P, keyof D>>>
+    & Partial<D>;
```
Comment (weswigham) [inline types/react/index.d.ts]: So, a note on this. If both `defaultProps` and `propTypes` are present, this implies that something while only appears as a `defaultProp` and not a `propType` is OK. If `P` is actually specified (because the component has type definitions, for example), this means that defaultProps can silently augment the declared props in the public interface of the component.

Is that desirable? Ehhhhhhh. The alternative isn't elegant tho - pass in separate types (or have two flavors of Defaultize) for declared props vs inferred props into `Defaultize`, and if declaredProps is a sentinel value (doesn't exist), and neither do inferred props, then you want this intersection member (so it becomes a primary inference source), otherwise you probably don't.

**Comment by ferdaber** [comment]
Kind of, I didn't want to open it against DefinitelyTyped/master because then the bot might start cc'ing all those maintainers again (it'll pick up the changes from the other branch).

File: types/react/index.d.ts (hunk 1 above)
Comment (ferdaber) [reply types/react/index.d.ts]: Yes, I think I'm going to change this whole thing too, in testing I found that naively intersecting the proper `Props` interface and the inferred `DefaultProps` will end up widening types for non-annotated prop types properties. Something like this is bad:
```tsx
interface Props {
  someEnum: 'a' | 'b' | 'c'
}
const propTypes = {
  someEnum: PropTypes.string.isRequired
}
class Foo extends Component<Props> {
  static propTypes = propTypes
}
const foo = <Foo /> // attributes['someEnum'] becomes type 'string'
```

TLDR; more to come! Gotta do more testing on some edge cases. I also *really* want it to work well with JS users because this has the potential to make JSX-based authoring so much better for VSCode (and other TS language server) users

File: types/react/index.d.ts [hunk 2]
```diff
@@ -2204,31 +2205,38 @@ declare namespace React {
...
+    type Requireable<T> = PropTypes.Requireable<T>;
 
-    interface Requireable<T> extends Validator<T> {
-        isRequired: Validator<T>;
-    }
+    type ValidationMap<T> = PropTypes.ValidationMap<T>;
 
-    type ValidationMap<T> = {[K in keyof T]?: Validator<T> };
+    // Infer a component's props from its static properties
+    type InferStaticProps<C> = C extends { propTypes: infer T; defaultProps: infer D; }
```
Comment (weswigham) [inline types/react/index.d.ts]: Is this used somewhere?

Comment (ferdaber) [reply types/react/index.d.ts]: Ah I think I left that in for some tests but later removed them, going to remove them now.

**Comment by ferdaber** [comment]
@weswigham if you can take a quick once-over for the React changes, that'd be appreciated. If you feel good about it I am going to move this PR to point to the DT repo, and have the bot ping 500 maintainers... 🙃

**Comment by weswigham** [comment]
@ferdaber I think we have a workaround for the issue we found on the other issue wrt. index signatures - I don't see this PR reopened on DT anywhere; or did I miss it?
